Crash report 3.1: ClassCastException in ReaderPostListFragment

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android/org.wordpress.android.ui.reader.ReaderPostListActivity}: java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.util.Stack
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2062)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2087)
       at android.app.ActivityThread.access$600(ActivityThread.java:133)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1198)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:146)
       at android.app.ActivityThread.main(ActivityThread.java:4777)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:511)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:789)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:556)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by: java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.util.Stack
       at org.wordpress.android.ui.reader.ReaderPostListFragment.onCreate(ReaderPostListFragment.java:185)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:796)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1035)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1017)
       at android.app.FragmentManagerImpl.dispatchCreate(FragmentManager.java:1800)
       at android.app.Activity.onCreate(Activity.java:886)
       at org.wordpress.android.ui.WPActionBarActivity.onCreate(WPActionBarActivity.java:112)
       at org.wordpress.android.ui.reader.ReaderPostListActivity.onCreate(ReaderPostListActivity.java:52)
       at android.app.Activity.performCreate(Activity.java:5008)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1079)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2026)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2087)
       at android.app.ActivityThread.access$600(ActivityThread.java:133)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1198)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:146)
       at android.app.ActivityThread.main(ActivityThread.java:4777)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:511)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:789)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:556)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

--- a/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java
@@ -56,6 +56,7 @@ import org.wordpress.android.util.ptr.PullToRefreshHelper;
 import org.wordpress.android.util.ptr.PullToRefreshHelper.RefreshListener;
 import org.wordpress.android.widgets.WPListView;
 
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Stack;
@@ -104,8 +105,30 @@ public class ReaderPostListFragment extends Fragment
 
     private Parcelable mListState = null;
 
-    private final Stack<String> mTagPreviewHistory = new Stack<String>();
-    private static final String KEY_TAG_PREVIEW_HISTORY = "tag_preview_history";
+    private final HistoryStack mTagPreviewHistory = new HistoryStack("tag_preview_history");
+
+    private static class HistoryStack extends Stack<String> {
+        private final String keyName;
+        HistoryStack(String keyName) {
+            this.keyName = keyName;
+        }
+        void restoreInstance(Bundle bundle) {
+            clear();
+            if (bundle.containsKey(keyName)) {
+                ArrayList<String> history = bundle.getStringArrayList(keyName);
+                if (history != null) {
+                    this.addAll(history);
+                }
+            }
+        }
+        void saveInstance(Bundle bundle) {
+            if (!isEmpty()) {
+                ArrayList<String> history = new ArrayList<String>();
+                history.addAll(this);
+                bundle.putStringArrayList(keyName, history);
+            }
+        }
+    }
 
     /*
      * show posts with a specific tag
@@ -181,10 +204,8 @@ public class ReaderPostListFragment extends Fragment
             if (savedInstanceState.containsKey(ReaderConstants.ARG_POST_LIST_TYPE)) {
                 mPostListType = (ReaderPostListType) savedInstanceState.getSerializable(ReaderConstants.ARG_POST_LIST_TYPE);
             }
-            if (savedInstanceState.containsKey(KEY_TAG_PREVIEW_HISTORY)) {
-                Stack<String> backStack = (Stack<String>) savedInstanceState.getSerializable(KEY_TAG_PREVIEW_HISTORY);
-                mTagPreviewHistory.clear();
-                mTagPreviewHistory.addAll(backStack);
+            if (getPostListType() == ReaderPostListType.TAG_PREVIEW) {
+                mTagPreviewHistory.restoreInstance(savedInstanceState);
             }
             mWasPaused = savedInstanceState.getBoolean(ReaderConstants.KEY_WAS_PAUSED);
         }
@@ -226,8 +247,8 @@ public class ReaderPostListFragment extends Fragment
         if (mCurrentTag != null) {
             outState.putSerializable(ReaderConstants.ARG_TAG, mCurrentTag);
         }
-        if (!mTagPreviewHistory.empty()) {
-            outState.putSerializable(KEY_TAG_PREVIEW_HISTORY, mTagPreviewHistory);
+        if (getPostListType() == ReaderPostListType.TAG_PREVIEW) {
+            mTagPreviewHistory.saveInstance(outState);
         }
 
         outState.putLong(ReaderConstants.ARG_BLOG_ID, mCurrentBlogId);

